Importing engine

In [2]:
import pandas as pd 
import sys 
import os 

sys.path.append(os.path.abspath("../scripts"))

from db_connection import engine

DB to python


In [7]:
table_names = ['customers','geolocation','order_items','order_payments', 'order_reviews','orders','products','product_category_name_translation','sellers']

tables = {}

for table in table_names:

    query = f"SELECT * FROM {table}"

    try:

        tables[table] = pd.read_sql(query,engine)

        print(f"The table {table} has been loaded successfully.")

    except Exception as error:
        print(f"The table {table} failed to load.")
        print(f"Reason: {error}")

The table customers has been loaded successfully.
The table geolocation has been loaded successfully.
The table order_items has been loaded successfully.
The table order_payments has been loaded successfully.
The table order_reviews has been loaded successfully.
The table orders has been loaded successfully.
The table products has been loaded successfully.
The table product_category_name_translation has been loaded successfully.
The table sellers has been loaded successfully.


validation

In [27]:
for table_name, df in tables.items():

    print("\n" + "=" * 70)
    print(f"{table_name.upper()}")
    print("=" * 70)

    print(f"\nShape: {df.shape}")

    print("\nData Types:")
    print(df.dtypes)

    # Missing values
    null_counts = df.isnull().sum()

    null_percentages = (
        null_counts / len(df)
    ) * 100

    null_summary = pd.DataFrame({
        "Null Count": null_counts,
        "Null Percentage": null_percentages.round(2)
    })

    print("\nMissing Values:")

    print(
        null_summary[
            null_summary["Null Count"] > 0
        ]
    )

    # Duplicate rows
    duplicate_count = df.duplicated().sum()

    duplicate_percentage = (
        duplicate_count / len(df)
    ) * 100

    print("\nDuplicate Rows:")
    print(duplicate_count)

    print("\nDuplicate Percentage:")
    print(f"{duplicate_percentage:.2f}%")



CUSTOMERS

Shape: (99441, 5)

Data Types:
customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object

Missing Values:
Empty DataFrame
Columns: [Null Count, Null Percentage]
Index: []

Duplicate Rows:
0

Duplicate Percentage:
0.00%

GEOLOCATION

Shape: (1000163, 5)

Data Types:
geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                   str
geolocation_state                  str
dtype: object

Missing Values:
Empty DataFrame
Columns: [Null Count, Null Percentage]
Index: []

Duplicate Rows:
262010

Duplicate Percentage:
26.20%

ORDER_ITEMS

Shape: (112650, 7)

Data Types:
order_id                          str
order_item_id                   int64
product_id                        str
seller_id                         str
shipping_limit_date    datetime64[us]
price          

Customers - Good

Geolocation - 26% duplicates --> look up dataset , no issue 

order_items - Good - 

order_payments - Good

order_reveiews - null/missing values ( comment title - 88% , message - 59%), duplicates(50%) -- not a huge problem/ product -- but the duplicates and missing values beats me  

orders - missing values(order_approved_at - 0.16%, delivered_carrrier_data - 1.79%, delivered_customer_date- 2.98%) slightly ok but we'get back to why 2.98% late deliveries.

products - (0.01 - 1.85%) ok 
